src/io/io_session.f90

```diff
@@ -0,0 +1,215 @@
+module m_io_session
+!! High-level I/O session providing a unified session-based interface
+!! This module abstracts away the complexity of the underlying I/O system,
+!! providing the io_session_t type for efficient multi-variable I/O operations
+!! and factory functions for creating the appropriate I/O backend.
+
+  use mpi, only: MPI_COMM_WORLD
+  use m_common, only: dp, i8
+  use m_io_base, only: io_reader_t, io_writer_t, io_file_t, &
+                       io_mode_read, io_mode_write
+  use m_io_factory, only: allocate_io_reader, allocate_io_writer
+
+  implicit none
+
+  private
+
+  ! Re-export factory functions through session module for proper layering
+  public :: allocate_io_reader, allocate_io_writer
+  ! Session-based API
+  public :: io_session_t
+
+  !> Type for reading and writing multiple variables from/to the same file efficiently
+  !!
+  !! Usage example for reading:
+  !!   type(io_session_t) :: io_session
+  !!   call io_session%open("checkpoint.bp", MPI_COMM_WORLD, io_mode_read)
+  !!   call io_session%read_data("timestep", timestep)
+  !!   call io_session%read_data("velocity_u", u_field, start_dims, count_dims)
+  !!   call io_session%close()
+  !!
+  !! Usage example for writing:
+  !!   type(io_session_t) :: io_session
+  !!   call io_session%open("output.bp", MPI_COMM_WORLD, io_mode_write)
+  !!   call io_session%write_data("timestep", current_step)
+  !!   call io_session%write_data("pressure", p_field, start_dims, count_dims)
+  !!   call io_session%write_attribute("ParaView", "vtk_xml_content")
+  !!   call io_session%close()
+  !!
+  !! The session automatically handles backend selection (currently just ADIOS2 but can be extended to MPI-IO, etc.)
+  !! and manages file handles, readers, and writers internally.
+  type :: io_session_t
+    private
+    class(io_reader_t), pointer :: reader => null()
+    class(io_writer_t), pointer :: writer => null()
```

Comment:
Do we want a single object to support reading and writing or is the intention that you should have a reader session and a writer session? If the latter would it be better to specialise the type?